File: red/src/workflow.ts
import {workflow, type Opts, type StepFn, type Workflow} from 'red/workflow';
import {collect, state_keys} from './index.ts';

/** Build a Colors fork/join around an injected single-node operation.
 * nodeStep receives colors-compute/request and returns colors-compute/params.
 * It must throw or return red/exit > 0 on failure. No provisioning is supplied
 * by this adapter; the caller supplies the common single-node lifecycle step.
 */
export function clusterWorkflow(
  requests: Opts[],
  entryNodeId: string,
  nodeStep: StepFn,
  downstream?: StepFn,
): Workflow {
  if (!requests.length) throw new Error('no nodes requested');
  const ids = new Set<string>();
  for (const request of requests) {
    if (ids.has(request.node_id)) throw new Error(`duplicate requested node: ${request.node_id}`);
    ids.add(request.node_id);
  }
  if (!ids.has(entryNodeId)) throw new Error(`unknown entry node: ${entryNodeId}`);
  state_keys('validation', requests.map(request => request.node_id));
  // Snapshot configuration so caller mutation cannot change a constructed DAG.
  const declared = structuredClone(requests);
  return workflow({
    start: 'colors-compute/fan-out',
    wireFn: (step) => {
      switch (step) {
        case 'colors-compute/fan-out':
          return [opts => ({...opts}), 'colors-compute/node'];
        case 'colors-compute/node':
          return [nodeStep, 'colors-compute/join'];
        case 'colors-compute/join':
          return [opts => {
            // A one-node route has no fork, so the SDK supplies its result
            // directly rather than creating red/branches.
            const branches: Opts[] = declared.length === 1 ? [opts] : opts['red/branches'];
            const results = branches.map(branch => branch['colors-compute/params'])
              .filter(result => result !== undefined && result !== null);
            return {...opts, 'colors-compute/cluster': collect(declared, results, entryNodeId)};
          }, ...(downstream ? ['colors-compute/downstream'] : [])];
        case 'colors-compute/downstream':
          return downstream ? [downstream] : undefined;
      }
    },
    nextFn: (step, next, opts) => {
      if ((opts['red/exit'] ?? 0) > 0) return [];
      if (step === 'colors-compute/fan-out') {
        return declared.map(request => ['colors-compute/node', {
          ...opts, 'colors-compute/request': structuredClone(request),
        }] as const);
      }
      return (next ?? []).map(name => [name, opts] as const);
    },
  });
}

File: blue/src/colors_compute/workflow.py
"""Colors SDK fan-out over the common node step, followed by a checked join.

This constructor composes supplied node operations. It does not itself prepare
keys, acquire remote deployment ownership, or execute OpenTofu.
"""

from __future__ import annotations

from ._copy import deepcopy
from typing import Callable
import inspect

from blue.workflow import Workflow, workflow

from .contract import collect, state_keys


def cluster_workflow(requests: list[dict], entry_node_id: str,
                     node_step: Callable, downstream: Callable | None = None) -> Workflow:
    if not requests:
        raise ValueError("no nodes requested")
    seen = set()
    for request in requests:
        node_id = request.get("node_id")
        if node_id in seen:
            raise ValueError(f"duplicate requested node: {node_id}")
        seen.add(node_id)
    state_keys("validation", [request.get("node_id") for request in requests])
    if entry_node_id not in seen:
        raise ValueError(f"unknown entry node: {entry_node_id}")
    if not callable(node_step):
        raise ValueError("node-step must be callable")
    declared = deepcopy(requests)

    def dispatch(opts):
        return {key: value for key, value in opts.items()
                if key not in ("colors-compute/cluster", "colors-compute/params", "blue/branches")}

    def join(opts):
        branches = opts.get("blue/branches") or [opts]
        result = collect(declared, [branch["colors-compute/params"] for branch in branches
                                    if branch.get("colors-compute/params") is not None], entry_node_id)
        return {**opts, "colors-compute/cluster": result}

    async def guarded_node(opts):
        result = node_step(opts)
        if inspect.isawaitable(result):
            result = await result
        code = result.get("blue/exit")
        if code is not None and not (type(code) in (int, float) and code >= 0):
            return {**result, "blue/exit": 1}
        return result

    def wire_fn(step, _run_opts):
        return {
            "colors-compute/dispatch": (dispatch, "colors-compute/node"),
            "colors-compute/node": (guarded_node, "colors-compute/join"),
            "colors-compute/join": ((join, "colors-compute/downstream") if downstream else (join,)),
            "colors-compute/downstream": (downstream,),
        }.get(step)

    def next_fn(step, successors, opts):
        if (opts.get("blue/exit") or 0) != 0:
            return []
        if step == "colors-compute/dispatch":
            return [("colors-compute/node", {**opts, "colors-compute/request": deepcopy(request)})
                    for request in declared]
        return [(successor, opts) for successor in successors or []]

    return workflow(start="colors-compute/dispatch", wire_fn=wire_fn, next_fn=next_fn)
